### DISCUSSION TITLE
How to trigger context menu to pull menu items each time user right clicks?

### DISCUSSION THREAD

**Discussion by cornerbowlsoftware** [AvaloniaUI/Avalonia#16315] [General]
I need to update my menu items each time the user right clicks on any control that has a context menu.  For TreeViews in WPF I handle it like so:

```
        private void Tvw_PreviewMouseRightButtonDown(object? sender, MouseButtonEventArgs e)
        {
            try
            {
                if (e.OriginalSource is DependencyObject depObj)
                {
                    var tvi = depObj.FindParent<TreeViewItem>();
                    if (tvi != null)
                    {
                        if (tvi.DataContext is TreeNodeViewModel vm)
                        {
                            tvi.IsSelected = true;
                            e.Handled = true;

                            _ = tvw.Focus();
                            vm.NotifyPropertyChanged(nameof(TreeNodeViewModel.MenuItemVisibility));
                            vm.NotifyPropertyChanged(nameof(TreeNodeViewModel.MenuItemVMs));
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Commands.DisplayException(ex);
            }
        }
```

Avalonia does not have mouse preview events and the pointerpressed event isn't called when clicking on a tree node within a treeview.  How can I implement dynamic context menu items?

**Comment by stevemonaco** [comment]
There aren't separate preview events, but you can register a tunnelling event from code-behind instead. Although you should be able to update what you want via bindings. For tunnelling events, see https://docs.avaloniaui.net/docs/get-started/wpf/tunnelling-events

It's a bit indirect to get the `TreeViewItem` associated with the click if you subscribe to `TreeView`. Subscribing to individual items is harder, IMO. You'll still need to bridge the gap with showing/updating the `ContextMenu` yourself.

```xml
<TreeView x:Name="tree">
    <TreeViewItem Header="Test A" />
    <TreeViewItem Header="Test B" />
    <TreeViewItem Header="Test C" />
</TreeView>
```

```cs
public partial class MainWindow : Window
{
    public MainWindow()
    {
        InitializeComponent();

        tree.AddHandler(PointerPressedEvent, Tree_PointerPressed, RoutingStrategies.Tunnel);
    }

    private void Tree_PointerPressed(object? sender, PointerPressedEventArgs e)
    {
        if (e.Pointer.Type != PointerType.Mouse)
            return;

        var point = e.GetCurrentPoint(tree);

        if (point.Properties.PointerUpdateKind == PointerUpdateKind.RightButtonPressed)
        {
            TreeViewItem tvi = ((Control)e.Source).GetVisualAncestors().OfType<TreeViewItem>().First();
            // Do work
            e.Handled = true;
        }
    }
}
```

**Reply by stevemonaco** [reply]
Also, this only works for mouse. So you probably want to try another way as users can launch context menus via keyboard, touch, etc.

**Comment by cornerbowlsoftware** [comment]
Thanks for the information.  That looks great!

**Reply by cornerbowlsoftware** [reply]
Also, since I didn't yet know about this, I ended up overriding 'CanExecute' in my Commands and that seems to do the trick without hiding potential menu items.

**Comment by maxkatz6** [comment]
Use ContextRequested event. It's more suitable than PointerPressed.

**Reply by maxkatz6** [reply]
It also will be executed, when context was requested with a keyboard, for example.
